tools/clang_format.py

```diff
@@ -0,0 +1,124 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+import argparse
+import logging
+import os
+import re
+import subprocess
+import sys
+import tempfile
+import typing
+
+import githelper
+
+
+# We recommend a maximum line length of 80, but do allow up to 100 characters
+# if deemed necessary by the developer. Lines that exceed that limit will
+# be wrapped after 80 characters automatically.
+LINE_LENGTH_THRESHOLD = 100
+BREAK_BEFORE = 80
+
+
+def get_clang_format_config_with_columnlimit(rootdir, limit):
+    cpp_file = os.path.join(rootdir, "src/mixxx.cpp")
+    proc = subprocess.run(
+        ["clang-format", "--dump-config", cpp_file],
+        capture_output=True,
+        text=True,
+    )
+    proc.check_returncode()
+    return re.sub(r"(ColumnLimit:\s*)\d+", r"\g<1>{}".format(80), proc.stdout,)
+
+
+def run_clang_format_on_lines(rootdir, changed_file, assume_filename=None):
+    logger = logging.getLogger(__name__)
+
+    line_arguments = [
+        "--lines={}:{}".format(start, end) for start, end in changed_file.lines
+    ]
+    assert line_arguments
+
+    logger.info("Reformatting %s...", changed_file.filename)
+    filename = os.path.join(rootdir, changed_file.filename)
+    cmd = [
+        "clang-format",
+        "--style=file",
+        "--assume-filename={}".format(
+            assume_filename if assume_filename else filename
+        ),
+        *line_arguments,
+    ]
+
+    with open(filename) as fp:
+        logger.debug("Executing: %r", cmd)
+        proc = subprocess.run(cmd, stdin=fp, capture_output=True, text=True)
+    try:
+        proc.check_returncode()
+    except subprocess.CalledProcessError:
+        logger.error(
+            "Error while executing command %s: %s", cmd, proc.stderr,
+        )
+        raise
+
+    if proc.stderr:
+        logger.error(proc.stderr)
+    with open(filename, mode="w+") as fp:
+        fp.write(proc.stdout)
+
+
+def main(argv: typing.Optional[typing.List[str]] = None) -> int:
+    logging.basicConfig(
+        format="[%(levelname)s] %(message)s", level=logging.INFO
+    )
+
+    logger = logging.getLogger(__name__)
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--from-ref", help="compare against git reference")
+    parser.add_argument("files", nargs="*", help="only check these files")
```

Comment:
Why add it if it only throws an error?

Reply:
Because it is part of the API. 
What happens if one give the argument "--to-ref" do we throw an exception anyway? 
What happens if one sets "PRE_COMMIT_TO_REF" 
In these cases the user explicit don't want to check against HEAD, If we silently do it anyway, we are IMHO doing wrong.

Reply:
Travis uses 
```
pre-commit run --from-ref HEAD^1 --to-ref HEAD --show-diff-on-failure
```

Reply:
> What happens if one give the argument "--to-ref" do we throw an exception anyway?

Yes, argparse will complain:

```
$ tools/clang_format_wrapper.py --to-ref HEAD
usage: clang_format_wrapper.py [-h] [files [files ...]]
clang_format_wrapper.py: error: unrecognized arguments: --to-ref
```

Reply:
I added support for to_ref now.